godot-core/src/builtin/vector_macros.rs

```diff
@@ -0,0 +1,221 @@
+                $(
+                    self.$components = self.$components.$func();
+                )*
+                self
+            }
+        }
+    }
+}
+
+/// Implements a single infix binary operator for a vector type, between two vectors and between a
+/// vector and a scalar in both directions.
+macro_rules! impl_vector_binary_operator {
+    (
+        // Name of the vector type.
+        $vector:ty,
+        // Type of each individual component, for example `i32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `Add`.
+        $operator:ident,
+        // Name of the function on the operator trait, for example `add`.
+        $func:ident$(,)?
+    ) => {
+        // vector + vector
+        impl std::ops::$operator for $vector {
+            type Output = Self;
+            fn $func(mut self, rhs: $vector) -> Self::Output {
+                $(
+                    self.$components = self.$components.$func(rhs.$components);
+                )*
+                self
+            }
+        }
+
+        // vector + scalar
+        impl std::ops::$operator<$scalar> for $vector {
+            type Output = Self;
+            fn $func(mut self, rhs: $scalar) -> Self::Output {
+                $(
+                    self.$components = self.$components.$func(rhs);
+                )*
+                self
+            }
+        }
+
+        // scalar + vector
+        impl std::ops::$operator<$vector> for $scalar {
+            type Output = $vector;
+            fn $func(self, mut rhs: $vector) -> Self::Output {
+                $(
+                    rhs.$components = rhs.$components.$func(self);
+                )*
+                rhs
+            }
+        }
+    }
+}
+
+/// Implements a single arithmetic assignment operator for a vector type.
+macro_rules! impl_vector_assign_operator {
+    (
+        // Name of the vector type.
+        $vector:ty,
+        // Type of each individual component, for example `i32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `AddAssign`.
+        $operator:ident,
+        // Name of the function on the operator trait, for example `add_assign`.
+        $func:ident$(,)?
+    ) => {
+        // vector += vector
+        impl std::ops::$operator for $vector {
+            fn $func(&mut self, rhs: $vector) {
+                $(
+                    self.$components.$func(rhs.$components);
+                )*
+            }
+        }
+
+        // vector += scalar
+        impl std::ops::$operator<$scalar> for $vector {
+            fn $func(&mut self, rhs: $scalar) {
+                $(
+                    self.$components.$func(rhs);
+                )*
+            }
+        }
+    }
```

Comment:
See above.

In general, the operation `a @= b` should be provided **if and only if** `a = a @ b` is provided.
(for any operator `@`)